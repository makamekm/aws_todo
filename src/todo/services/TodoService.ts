import { useMutation, useQuery } from "graphql-hooks";
import * as moment from "moment";
import * as React from "react";
import { inject } from "react-ioc";
import { BehaviorSubject, from, Subject } from "rxjs";
import { useObservable } from "rxjs-hooks";
import { map, switchMap, tap, toArray } from "rxjs/operators";
import { StoreService } from "../../iso/services/StoreService";

export const QUERY_TODO_LIST = `query ListTodo($year: Int!, $month: Int!, $day: Int!) {
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;
export const QUERY_TODO_STATS = `query StatsTodo($year: Int!, $month: Int!) {
    todoQuery {
        stats(
            year: $year,
            month: $month,
        ) {
            year
            month
            day
            finished
            total
            unfinished {
                id
                name
                isDone
            }
        }
    }
}`;

export const MUTATION_TODO_CREATE = `mutation CreateTodo(
        $year: Int!,
        $month: Int!,
        $day: Int!,
        $name: String!
    ) {
    todoMutation {
        create(
            year: $year,
            month: $month,
            day: $day,
            name: $name,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_DELETE = `mutation DeleteTodo(
        $id: ID!,
        $year: Int!,
        $month: Int!,
        $day: Int!,
    ) {
    todoMutation {
        delete(id: $id) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_TOGGLE = `mutation ToggleTodo(
    $id: ID!,
    $year: Int!,
    $month: Int!,
    $day: Int!,
) {
    todoMutation {
        toggle(id: $id) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_EDIT = `mutation EditTodo(
        $id: ID!,
        $name: String!,
        $year: Int!,
        $month: Int!,
        $day: Int!,
    ) {
    todoMutation {
        edit(id: $id, name: $name) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export interface ITodo {
    id: number;
    name: string;
    isDone: boolean;
    date: string;
    updated: string;
}

export class TodoService {
    @inject public storeService: StoreService;
    public searchString$ = new BehaviorSubject("");
    public data$: BehaviorSubject<ITodo[]> = new BehaviorSubject(null);
    public mutationError$: Subject<string> = new Subject();
    public queryErrors$: BehaviorSubject<string[]> = new BehaviorSubject([]);
    public date$ = new BehaviorSubject(moment(this.storeService.date));

    public findStats(stats, year, month, day) {
        return stats && stats.find((s) => s.year === year && s.month === month && s.day === day);
    }

    public useStatsService = () => {
        const response = useQuery(QUERY_TODO_STATS, {
            variables: {
                year: this.date$.value.year(),
                month: this.date$.value.month(),
            },
        });
        const { data: queryData, loading, refetch, error } = response;
        const stats = queryData && queryData.todoQuery && queryData.todoQuery.stats;
        useObservable(() => this.date$.pipe(
            switchMap(refetch),
        ));

        if (error) {
            this.pipeQueryErrors(response);
        }

        return {
            stats,
            loading,
            refetch,
            getStats: (date) => this.findStats(
                stats,
                date.year(),
                date.month(),
                date.date(),
            ),
        };
    }

    public useDataService = () => {
        const response = useQuery(QUERY_TODO_LIST, {
            variables: {
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        const { data: queryData, loading, refetch, error } = response;
        const list = queryData && queryData.todoQuery && queryData.todoQuery.list;

        if (error) {
            this.pipeQueryErrors(response);
        }

        if (!this.data$.value && list) {
            this.data$.next(list);
        }

        React.useEffect(() => {
            if (list) {
                this.data$.next(list);
            }
        }, [list]);

        return {
            refetch,
            loading,
        };
    }

    public useItemService = (item) => {
        const [deleteTodo, { loading: deleteTodoLoading }] = useMutation(MUTATION_TODO_DELETE);
        const [toggleTodo, { loading: toggleTodoLoading }] = useMutation(MUTATION_TODO_TOGGLE);
        const [editTodo, { loading: editTodoLoading }] = useMutation(MUTATION_TODO_EDIT);

        return {
            loading: deleteTodoLoading || toggleTodoLoading || editTodoLoading,
            deleteTodo: () => this.deleteTodo(deleteTodo, item.id),
            toggleTodo: () => this.toggleTodo(toggleTodo, item.id),
            editTodo: (name) => this.editTodo(editTodo, item.id, name),
        };
    }

    public useCreateService = () => {
        const [createTodo, { loading: createTodoLoading }] = useMutation(MUTATION_TODO_CREATE);

        return {
            loading: createTodoLoading,
            createTodo: () => this.createTodo(createTodo, this.searchString$.value),
        };
    }

    public removeQueryError(index: number) {
        const list = Array.from(this.queryErrors$.value);
        list.splice(index, 1);
        this.queryErrors$.next(list);
    }

    private pipeMutationErrors(response) {
        from(response.graphQLErrors).pipe(
            map((error: any) => error.message),
            tap((error: string) => this.mutationError$.next(error)),
        ).subscribe();
    }

    private pipeQueryErrors(response) {
        from(response.graphQLErrors).pipe(
            map((error: any) => error.message),
            toArray(),
            tap((errors: string[]) => this.queryErrors$.next(errors)),
        ).subscribe();
    }

    private async createTodo(createTodo, name) {
        const result = await createTodo({
            variables: {
                name,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async deleteTodo(deleteTodo, id) {
        const result = await deleteTodo({
            variables: {
                id,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async editTodo(editTodo, id, name) {
        const result = await editTodo({
            variables: {
                id,
                name,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async toggleTodo(toggleTodo, id) {
        const result = await toggleTodo({
            variables: {
                id,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }
}
