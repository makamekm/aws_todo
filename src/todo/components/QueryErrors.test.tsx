import * as React from "react";
import {cleanup} from "react-testing-library";
import { getTodoService, renderAndWrapWithService } from "../test/TodoService";
import {
    QueryErrors,
} from "./QueryErrors";

describe("Todo", () => {
    describe("Components", () => {
        describe("QueryErrors", () => {
            afterEach(cleanup);

            it("should match snapshot", async () => {
                const service = getTodoService();
                service.queryErrors$.next(["test1", "test2"]);
                const html = await renderAndWrapWithService(
                    <QueryErrors/>,
                    service,
                );
                expect(html.container).toMatchSnapshot();
            });
        });
    });
});
