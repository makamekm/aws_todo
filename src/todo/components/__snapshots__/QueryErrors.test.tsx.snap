// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Components QueryErrors should match snapshot 1`] = `
<div>
  <div
    class="ant-alert ant-alert-error ant-alert-closable"
    data-show="true"
    style="margin-bottom: 5px;"
  >
    <i
      aria-label="icon: close-circle"
      class="anticon anticon-close-circle ant-alert-icon"
    >
      <svg
        aria-hidden="true"
        class=""
        data-icon="close-circle"
        fill="currentColor"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
        />
      </svg>
    </i>
    <span
      class="ant-alert-message"
    >
      test1
    </span>
    <span
      class="ant-alert-description"
    />
    <a
      class="ant-alert-close-icon"
    >
      <i
        aria-label="icon: close"
        class="anticon anticon-close"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="close"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
          />
        </svg>
      </i>
    </a>
  </div>
  <div
    class="ant-alert ant-alert-error ant-alert-closable"
    data-show="true"
    style="margin-bottom: 5px;"
  >
    <i
      aria-label="icon: close-circle"
      class="anticon anticon-close-circle ant-alert-icon"
    >
      <svg
        aria-hidden="true"
        class=""
        data-icon="close-circle"
        fill="currentColor"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
        />
      </svg>
    </i>
    <span
      class="ant-alert-message"
    >
      test2
    </span>
    <span
      class="ant-alert-description"
    />
    <a
      class="ant-alert-close-icon"
    >
      <i
        aria-label="icon: close"
        class="anticon anticon-close"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="close"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
          />
        </svg>
      </i>
    </a>
  </div>
</div>
`;
